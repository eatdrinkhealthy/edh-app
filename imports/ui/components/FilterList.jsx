// @flow
import React, { Component } from "react";
import { Link } from "react-router";
import Toggle from "react-toggle";
import type { IFilter } from "../../data/state/data/defaultFilters";

type IFilterItemProps = {
  label: string,
  filterId: string,
  filterOn: boolean,
  setFilterHandler: (id: string, checked: boolean) => void,
};

export class FilterItem extends Component {
  handleToggleChange(event: Event) {
    const target = (event.target: window.HTMLInputElement);  // TODO issue #37
    this.props.setFilterHandler(target.id, target.checked);
  }

  props: IFilterItemProps;

  render() {  // eslint-disable-line flowtype/require-return-type
    return (
      <div>
        <span className="react-toggle-label">{this.props.label}</span>
        <span className="react-toggle">
          <Toggle
            id={this.props.filterId}
            checked={this.props.filterOn}
            onChange={(event: Event): void => this.handleToggleChange(event)}
          />
        </span>
      </div>
    );
  }
}

export type IFilterListProps = {
  filterList?: Array<IFilter>,
  setFilterHandler: () => void,
};

const FilterList = (
  { filterList = [], setFilterHandler }: IFilterListProps,
): React$Element<*> => (
  <div>
    <div className="filter-header">
      <h4>Filter</h4>
      <Link to="/">
        <div className="close-filter toggle-filter" />
      </Link>
    </div>
    <div className="filter">
      {filterList.map((filter: IFilter): React.Element<*> =>
        (<FilterItem
          key={filter.id}
          label={filter.name}
          filterId={filter.id}
          filterOn={filter.on}
          setFilterHandler={setFilterHandler}
        />))}
    </div>
  </div>
);

export default FilterList;
