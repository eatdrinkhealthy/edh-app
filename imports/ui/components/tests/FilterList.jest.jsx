// @flow
/* eslint-env jest */
/* eslint-disable func-names, prefer-arrow-callback, no-unused-expressions */
/* eslint-disable import/no-extraneous-dependencies */
import React from "react";
import renderer from "react-test-renderer";
import {
  shallow,
  mount,
} from "enzyme";
import FilterList, { FilterItem } from "../FilterList";
import type { IFilterList } from "../../../data/state/data/defaultFiltersTypes";

describe("Filter List Components", function () {
  const testFilterList: IFilterList = [
    {
      id: "juiceBar1",
      name: "Juice Bar1",
      on: true,
      foursquareCategory: "1",
    },
    {
      id: "juiceBar2",
      name: "Juice Bar2",
      on: false,
      foursquareCategory: "2",
    },
    {
      id: "juiceBar3",
      name: "Juice Bar3",
      on: false,
      foursquareCategory: "3",
    },
  ];

  describe("<FilterItem />", function () {
    it("matches render snapshot", function () {
      const tree = renderer.create(
        <FilterItem
          label={testFilterList[0].name}
          filterId={testFilterList[0].id}
          filterOn={testFilterList[0].on}
          setFilterHandler={jest.fn()}
        />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it("should call setFilter when clicked", function () {
      const props = {
        label: testFilterList[0].name,
        filterId: testFilterList[0].id,
        filterOn: testFilterList[0].on,
        setFilterHandler: jest.fn(),
      };

      const wrapper = shallow(<FilterItem {...props} />);
      const toggleComponent = wrapper.find("Toggle");
      expect(toggleComponent.length).toEqual(1);

      const event = {
        target: {
          id: "abc",
          checked: true,
        },
      };

      toggleComponent.props().onChange(event);
      expect(props.setFilterHandler.mock.calls.length).toBe(1);
    });
  });

  describe("<FilterList />", function () {
    it("matches render snapshot, with a filter list", function () {
      const tree = renderer.create(
        <FilterList
          filterList={testFilterList}
          setFilterHandler={jest.fn()}
        />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it("matches render snapshot, with no filter list", function () {
      const tree = renderer.create(
        <FilterList
          setFilterHandler={jest.fn()}
        />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it("renders a toggle component for each filter in the list", function () {
      const wrapper = mount(
        <FilterList
          filterList={testFilterList}
          setFilterHandler={jest.fn()}
        />);

      expect(wrapper.find("Toggle").length).toBe(3);
    });

    it("renders no toggle components when no filter list provided", function () {
      const wrapper = mount(
        <FilterList
          setFilterHandler={jest.fn()}
        />);

      expect(wrapper.find("Toggle").length).toBe(0);
    });
  });
});
