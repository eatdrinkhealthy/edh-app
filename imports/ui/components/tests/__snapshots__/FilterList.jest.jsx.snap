exports[`Filter List Components <FilterItem /> matches render snapshot 1`] = `
<div>
  <span
    className="react-toggle-label">
    Juice Bar1
  </span>
  <span
    className="react-toggle">
    <div
      className="react-toggle react-toggle--checked"
      onClick={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}>
      <div
        className="react-toggle-track">
        <div
          className="react-toggle-track-check">
          <svg
            height="11"
            viewBox="0 0 14 11"
            width="14">
            <title>
              switch-check
            </title>
            <path
              d="M11.264 0L5.26 6.004 2.103 2.847 0 4.95l5.26 5.26 8.108-8.107L11.264 0"
              fill="#fff"
              fillRule="evenodd" />
          </svg>
        </div>
        <div
          className="react-toggle-track-x">
          <svg
            height="10"
            viewBox="0 0 10 10"
            width="10">
            <title>
              switch-x
            </title>
            <path
              d="M9.9 2.12L7.78 0 4.95 2.828 2.12 0 0 2.12l2.83 2.83L0 7.776 2.123 9.9 4.95 7.07 7.78 9.9 9.9 7.776 7.072 4.95 9.9 2.12"
              fill="#fff"
              fillRule="evenodd" />
          </svg>
        </div>
      </div>
      <div
        className="react-toggle-thumb" />
      <input
        checked={true}
        className="react-toggle-screenreader-only"
        id="juiceBar1"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        type="checkbox" />
    </div>
  </span>
</div>
`;

exports[`Filter List Components <FilterList /> matches render snapshot, with a filter list 1`] = `
<div>
  <div
    className="filter-header">
    <h4>
      Filter
    </h4>
    <a
      onClick={[Function]}
      style={Object {}}>
      <div
        className="close-filter toggle-filter" />
    </a>
  </div>
  <div
    className="filter">
    <div>
      <span
        className="react-toggle-label">
        Juice Bar1
      </span>
      <span
        className="react-toggle">
        <div
          className="react-toggle react-toggle--checked"
          onClick={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}>
          <div
            className="react-toggle-track">
            <div
              className="react-toggle-track-check">
              <svg
                height="11"
                viewBox="0 0 14 11"
                width="14">
                <title>
                  switch-check
                </title>
                <path
                  d="M11.264 0L5.26 6.004 2.103 2.847 0 4.95l5.26 5.26 8.108-8.107L11.264 0"
                  fill="#fff"
                  fillRule="evenodd" />
              </svg>
            </div>
            <div
              className="react-toggle-track-x">
              <svg
                height="10"
                viewBox="0 0 10 10"
                width="10">
                <title>
                  switch-x
                </title>
                <path
                  d="M9.9 2.12L7.78 0 4.95 2.828 2.12 0 0 2.12l2.83 2.83L0 7.776 2.123 9.9 4.95 7.07 7.78 9.9 9.9 7.776 7.072 4.95 9.9 2.12"
                  fill="#fff"
                  fillRule="evenodd" />
              </svg>
            </div>
          </div>
          <div
            className="react-toggle-thumb" />
          <input
            checked={true}
            className="react-toggle-screenreader-only"
            id="juiceBar1"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            type="checkbox" />
        </div>
      </span>
    </div>
    <div>
      <span
        className="react-toggle-label">
        Juice Bar2
      </span>
      <span
        className="react-toggle">
        <div
          className="react-toggle"
          onClick={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}>
          <div
            className="react-toggle-track">
            <div
              className="react-toggle-track-check">
              <svg
                height="11"
                viewBox="0 0 14 11"
                width="14">
                <title>
                  switch-check
                </title>
                <path
                  d="M11.264 0L5.26 6.004 2.103 2.847 0 4.95l5.26 5.26 8.108-8.107L11.264 0"
                  fill="#fff"
                  fillRule="evenodd" />
              </svg>
            </div>
            <div
              className="react-toggle-track-x">
              <svg
                height="10"
                viewBox="0 0 10 10"
                width="10">
                <title>
                  switch-x
                </title>
                <path
                  d="M9.9 2.12L7.78 0 4.95 2.828 2.12 0 0 2.12l2.83 2.83L0 7.776 2.123 9.9 4.95 7.07 7.78 9.9 9.9 7.776 7.072 4.95 9.9 2.12"
                  fill="#fff"
                  fillRule="evenodd" />
              </svg>
            </div>
          </div>
          <div
            className="react-toggle-thumb" />
          <input
            checked={false}
            className="react-toggle-screenreader-only"
            id="juiceBar2"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            type="checkbox" />
        </div>
      </span>
    </div>
    <div>
      <span
        className="react-toggle-label">
        Juice Bar3
      </span>
      <span
        className="react-toggle">
        <div
          className="react-toggle"
          onClick={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}>
          <div
            className="react-toggle-track">
            <div
              className="react-toggle-track-check">
              <svg
                height="11"
                viewBox="0 0 14 11"
                width="14">
                <title>
                  switch-check
                </title>
                <path
                  d="M11.264 0L5.26 6.004 2.103 2.847 0 4.95l5.26 5.26 8.108-8.107L11.264 0"
                  fill="#fff"
                  fillRule="evenodd" />
              </svg>
            </div>
            <div
              className="react-toggle-track-x">
              <svg
                height="10"
                viewBox="0 0 10 10"
                width="10">
                <title>
                  switch-x
                </title>
                <path
                  d="M9.9 2.12L7.78 0 4.95 2.828 2.12 0 0 2.12l2.83 2.83L0 7.776 2.123 9.9 4.95 7.07 7.78 9.9 9.9 7.776 7.072 4.95 9.9 2.12"
                  fill="#fff"
                  fillRule="evenodd" />
              </svg>
            </div>
          </div>
          <div
            className="react-toggle-thumb" />
          <input
            checked={false}
            className="react-toggle-screenreader-only"
            id="juiceBar3"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            type="checkbox" />
        </div>
      </span>
    </div>
  </div>
</div>
`;

exports[`Filter List Components <FilterList /> matches render snapshot, with no filter list 1`] = `
<div>
  <div
    className="filter-header">
    <h4>
      Filter
    </h4>
    <a
      onClick={[Function]}
      style={Object {}}>
      <div
        className="close-filter toggle-filter" />
    </a>
  </div>
  <div
    className="filter" />
</div>
`;
